framework/src/service/ServiceReferenceBasePrivate.cpp

```diff
@@ -103,64 +114,58 @@ InterfaceMapConstPtr ServiceReferenceBasePrivate::GetPrototypeService(const Bund
 
 std::shared_ptr<void> ServiceReferenceBasePrivate::GetService(BundlePrivate* bundle)
 {
-  auto s = ExtractInterface(GetServiceInterfaceMap(bundle), interfaceId);
-  if (!s)
-  {
-    registration->Lock(), --registration->dependents[bundle];
-  }
-  return s;
+  return ExtractInterface(GetServiceInterfaceMap(bundle), interfaceId);
 }
 
 InterfaceMapConstPtr ServiceReferenceBasePrivate::GetServiceInterfaceMap(BundlePrivate* bundle)
 {
   InterfaceMapConstPtr s;
   if (!registration->available) return s;
-
   std::shared_ptr<ServiceFactory> serviceFactory;
-  int count = 0;
 
   {
     auto l = registration->Lock(); US_UNUSED(l);
     if (!registration->available) return s;
     serviceFactory = std::static_pointer_cast<ServiceFactory>(
           registration->GetService_unlocked("org.cppmicroservices.factory"));
-    count = registration->dependents[bundle];
-  }
 
-  if (!serviceFactory)
-  {
-    auto l = registration->Lock(); US_UNUSED(l);
-          s = registration->service;
-    if (s && !s->empty()) ++registration->dependents[bundle];
-  }
-  else
-  {
-    if (count == 0)
-    {
-      s = GetServiceFromFactory(bundle, serviceFactory);
+    registration->dependents.insert(std::make_pair(bundle, 0));
 
-      auto l = registration->Lock(); US_UNUSED(l);
-      if (registration->dependents[bundle] == 0)
-      {
-        registration->bundleServiceInstance.insert(std::make_pair(bundle, s));
-      }
-      else
+    if (!serviceFactory)
+    {
+      s = registration->service;
+      if (s && !s->empty())
       {
-        // There was a race and we now have one instance too much. Return the
-        // already produced instance and ignore the additional one.
-        s = registration->bundleServiceInstance[bundle];
+        ++registration->dependents.at(bundle);
       }
-
-      if (s && !s->empty()) ++registration->dependents[bundle];
+      return s;
     }
-    else
+
+    if (registration->bundleServiceInstance.end() != registration->bundleServiceInstance.find(bundle))
     {
-      auto l = registration->Lock(); US_UNUSED(l);
-      // return the already produced instance
-      s = registration->bundleServiceInstance[bundle];
-      if (s && !s->empty()) ++registration->dependents[bundle];
+      s = registration->bundleServiceInstance.at(bundle);
+      ++registration->dependents.at(bundle);
+      return s;
     }
   }
+
+  // Calling into a service factory could cause re-entrancy into the
+  // framework and even, theoretically, into this function. Ensuring
+  // we don't hold a lock while calling into the service factory eliminates
+  // the possibility of a deadlock. It does not however eliminate the
+  // possbilituy of infinite recursion.
+  s = GetServiceFromFactory(bundle, serviceFactory);
+
+  auto l = registration->Lock(); US_UNUSED(l);
+
+  registration->dependents.insert(std::make_pair(bundle, 0));
+
+  if (s && !s->empty())
```

Comment:
I agree, it is a strange yet possible scenario. 

If one thread has a null service object and another has a valid service object, what happens all depends on which thread enters the lock first.

If a service factory does return a null service object, I don't feel its the Framework's responsibility to try and guarantee that a valid one is always returned (excluding cases where one exists in the cache already). If a client's service factory returns a null service object on one thread and then a valid one on another, that seems like a bug; either with the client code querying for the service object or the client's service factory implementation. 

I'm not sure how the else condition really helps ensure that a non-null service is returned. This feels like we're overcompensating for possible bugs in client code. To check in an else branch we still have to hold the lock.